Implemented Trucated model

evgenyzhurko opened the pull request.

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,325 @@
+        """
+        llf_main = self.model_main.loglikeobs(params)
+
+        pmf = np.zeros_like(self.endog)
```
dtype in case endog is int

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,325 @@
+        for i in range(self.trunc + 1):
+            model = self.model_main_name(np.ones_like(self.endog) * i,
+                                         self.exog)
+            pmf +=  np.exp(model.loglikeobs(params))
```
this loop could be put into a `cdf` method that can be overwritten by a subclass.
E.g. poisson should have a build in method (or method delegated to scipy distribution cdf)

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,325 @@
+                                         self.exog)
+            pmf_i =  np.exp(model.loglikeobs(params))
+            score_trunc += (model.score_obs(params).T * pmf_i).T
+            pmf += pmf_i
```
I'm not sure whether it will be very useful, but this could be put into a separate cdf_deriv method.

Overall, for the design: we should outsource parts to separate methods, if it is useful to override the method for specific models or distributions that have a analytical or computational more efficient way to calculate those parts.

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,325 @@
+                                         self.exog)
+            pmf_i =  np.exp(model.loglikeobs(params))
+            score_trunc += (model.score_obs(params).T * pmf_i).T
+            pmf += pmf_i
```
? now this is cdf and deriv at the same time. 
So maybe one method that optionally returns either or both.

(but then there is still hessian_trunc)

evgenyzhurko commented:
@josef-pkt I tried to evaluate generic hessian, but it isn't successfully. The main problem that truncated model have variable amount of pmf(depends of truncation value)
llf = pmf_main(y, x) / (1 - pmf_main(0, x)) for Zero-Truncated
llf = pmf_main(y, x) / (1 - pmf_main(1, x) + pmf_main(2, x)) for truncation=2 and so on.
I skip analytic hessian for now.
I'll return for analytic hessian later.

For now, model works well for different trucation(i tested 0 and 5), results equal to STATA.
This model will be a good base for altered(adjusted) model.

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,355 @@
+        elif which == 'linear':
+            return linpred
+        elif which =='prob':
+            counts = np.atleast_2d(np.arange(0, np.max(self.endog)+1))
```
I think starting at zero is good for consistency, even though probabilities below `trunc` will be zero.

I think we should add an optional `count_prob` or similarly named argument for this, in analogy to the `n` parameter in PoissonResults.predict_prob

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,355 @@
+
+    """ + base._missing_param_doc}
+
+    def __init__(self, endog, exog, truncation=0, offset=None,
```
truncation params needs to be added to the docstring.
Specifically, it needs to specify whether the truncation point is in the support or out of support.
i.e. pmf(k) = 0 for k < truncation or for k<= truncation

josef-pkt commented:
doc strings still need work, several are incomplete

josef-pkt commented:
travis test runs fail on python 3 because of a left in py 2 print statement
```
File "/home/travis/build/statsmodels/statsmodels/statsmodels/discrete/count_model.py", line 353
    print endog[endog > 0].mean()
```

bashtage commented on statsmodels/discrete/tests/test_count_model.py:
```diff
@@ -0,0 +1,72 @@
+
+class TestTruncatedPoissonModel(object):
+    @classmethod
+    def setupClass(cls):
```
please use `setup_class` to simplfy moving to pytest.

bashtage commented on statsmodels/discrete/tests/test_count_model.py:
```diff
@@ -0,0 +1,72 @@
+
+class TestZeroTruncatedPoissonModel(object):
+    @classmethod
+    def setupClass(cls):
```
Same here

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,534 @@
+
+    """ + base._missing_param_doc}
+
+    def __init__(self, endog, exog, offset=None,
```
I think we should have two exog as in zero-inflated, one for the zeros and one for y > 0

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,534 @@
+
+        y = self.endog
+        zero_idx = np.nonzero(y == 0)[0]
+        nonzero_idx = np.nonzero(y)[0]
```
to avoid recomputing this, it would be better to put this in `__init__` and attach as attributes.

It might be more efficient to also separate endog and exog into the zero and non-zero parts and attach as internal attributes. It would use more permanent memory, but might be faster. (maybe some timing is necessary to see whether it's worth it.)  If we have time before the end of GSOC we could some performance checks to see whether and where the speed can be improved, but for now the priority is to get a working and unit tested version.

josef-pkt replied:
actually, the split would be part of the models, since exog don't show up in these methods directly

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,534 @@
+                                            missing=missing, **kwargs)
+        self.model1 = self.model_name1(
+            np.zeros_like(self.endog, dtype=np.float64), self.exog)
+        self.model2 = self.model_name2(self.endog, self.exog)
```
shouldn't model2, the truncated model, only get endog and exog for non-zero endog observations (y > 0)?
y = 0 should not be an observation in the support of the zero-truncated model.
or is this not imposed at this point, i.e. in the generic setting?

evgenyzhurko commented:
@josef-pkt 
I refactored hurdle model, but i don't know how i can implement result class for 2 models.
I also implemented zero-censored model in additional to hurdle model, as a result hurdle model have a simple structure.

Do you have any idea how i can implement result class?

josef-pkt commented:
This pull request introduces 32 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-660783fe95f4e60f4746fb08531e912dc0b24c66)

**new alerts:**

* 29 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

evgenyzhurko commented:
@josef-pkt 
One more question in this PR:
Do you have any idea how i can print 2 result class as one?

Hurdle model consist from Truncated and Censored models, llf = sum of llf. Results of my hurdle model equals to R results.

josef-pkt commented:
@evgenyzhurko  I don't see what the Hurdle model is doing here, it doesn't have a fit method. There is now a Censored model, which I don't understand based on a quick skim.
Also there are no unit tests or examples.

Can you provide an example (e.g. put script in comment or notebook in a gist) how the Hurdle model is supposed to work currently?

evgenyzhurko commented:
@josef-pkt 
R results for comparison:
```
> summary(hurdle(mdvis ~ 1 + lncoins + idp, randhie, dist="poisson", zero.dist = "poisson"))

Call:
hurdle(formula = mdvis ~ 1 + lncoins + idp, data = randhie, 
    dist = "poisson", zero.dist = "poisson")

Pearson residuals:
    Min      1Q  Median      3Q     Max 
-1.3188 -1.0334 -0.5500  0.2187 28.7939 

Count model coefficients (truncated poisson with log link):
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)  1.482417   0.006298 235.378   <2e-16 ***
lncoins     -0.031407   0.002299 -13.661   <2e-16 ***
idp         -0.103559   0.010584  -9.785   <2e-16 ***
Zero hurdle model coefficients (censored poisson with log link):
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.383840   0.014259   26.92   <2e-16 ***
lncoins     -0.080955   0.004786  -16.92   <2e-16 ***
idp         -0.330639   0.021865  -15.12   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 

Number of iterations in BFGS optimization: 24 
Log-likelihood: -57000.63 on 6 Df
```
Code for example:
```
        self.model_name1 = Poisson
        self.model_name2 = Poisson
        self.model1 = Censored(self.endog, self.exog, model=self.model_name1).fit()
        self.model2 = Truncated(self.endog, self.exog, model=self.model_name2).fit()

        print(self.model1.params)
        print(self.model2.params)

        print(self.model2.llf + self.model1.llf)
```
And results:
```
Optimization terminated successfully.
         Current function value: 0.610932
         Iterations 5
Optimization terminated successfully.
         Current function value: 3.217542
         Iterations 4
[-0.08095491 -0.33063876  0.38383968]
[-0.03140711 -0.10355922  1.4824171 ]
-57000.627142
```
And summary:
We can avoid complex computations in Hurdle model. We can unite 2 models as one and it ill work well.

evgenyzhurko commented:
@josef-pkt 
Last results:
My:
```
                       HurdlePoisson Regression Results
==============================================================================
Dep. Variable:                      y   No. Observations:                20190
Model:                  HurdlePoisson   Df Residuals:                    40374
Method:                           MLE   Df Model:                            4
Date:                Wed, 16 Aug 2017   Pseudo R-squ.:               5.700e+04
Time:                        16:47:23   Log-Likelihood:                -57001.
converged:               [True, True]   LL-Null:                        1.0000
                                        LLR p-value:                     1.000
=================================================================================
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
zero_x1          -0.0810      0.005    -16.916      0.000      -0.090      -0.072
zero_x2          -0.3306      0.022    -15.122      0.000      -0.373      -0.288
zero_const     0.3838      0.014     26.920      0.000       0.356       0.412
x1               -0.0314      0.002    -13.661      0.000      -0.036      -0.027
x2               -0.1036      0.011     -9.785      0.000      -0.124      -0.083
const             1.4824      0.006    235.378      0.000       1.470       1.495
=================================================================================
```
R
```

Pearson residuals:
    Min      1Q  Median      3Q     Max 
-1.3188 -1.0334 -0.5500  0.2187 28.7939 

Count model coefficients (truncated poisson with log link):
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)  1.482417   0.006298 235.378   <2e-16 ***
lncoins     -0.031407   0.002299 -13.661   <2e-16 ***
idp         -0.103559   0.010584  -9.785   <2e-16 ***
Zero hurdle model coefficients (censored poisson with log link):
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.383840   0.014259   26.92   <2e-16 ***
lncoins     -0.080955   0.004786  -16.92   <2e-16 ***
idp         -0.330639   0.021865  -15.12   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 

Number of iterations in BFGS optimization: 24 
Log-likelihood: -5.7e+04 on 6 Df
```
I'll try to finish result class refactoring in a short time.

josef-pkt commented:
This pull request introduces 33 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-bcd5006589d1d683cd95acf839acd244148c1f22)

**new alerts:**

* 29 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization
* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request introduces 34 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-d48c9ee0357535a19230693e076fce9b168c907c)

**new alerts:**

* 30 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization
* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request introduces 34 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-adc7b3f479a6e612884d1873b1c0cade673f1df1)

**new alerts:**

* 30 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization
* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request introduces 37 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-ffee7ce19018c158d0f914997f7f518f4e452023)

**new alerts:**

* 32 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization
* 1 for Syntax error
* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request introduces 34 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-3d4feadc1bf9b2221db7aa2630b372485afa6b10)

**new alerts:**

* 30 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization
* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request introduces 33 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-c627c227fcb7519ce48e7337900067627f42f5b8)

**new alerts:**

* 30 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request introduces 38 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-0c5b2a8d68537cb5bdacb60c730e32dff2950b50)

**new alerts:**

* 35 for Module-level cyclic import
* 2 for First argument to super() is not enclosing class
* 1 for Missing call to __init__ during object initialization

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
@evgenyzhurko  I don't think I have time to go through this in detail within the next few weeks (if we want to get the release out).

main question: What did you use as references for this, especially for the censored model and the hurdle model implemented through censored + truncated?

I just found an older Stata package http://www.stata-journal.com/article.html?article=st0219 and just realised that Stata 14 includes `cpoisson` as new model.
Based on the loglikeobs, the current version is not a general censored model, but only supports the model for the zeros as in the zero-modified hurdle model. (i.e. no censoring at user defined points larger than 1.) Given that, it needs to be either extended to arbitrary right censoring, or privatised for now, prepend underscore to the class names `_CensordXxx`, so it's clear that it is used only to support the hurdle models.
(I never looked at the literature specific to censored count models.)

evgenyzhurko commented:
@josef-pkt 
I have not got any articles for the Censored model, i looked through the hurdle model in the R and implemented in the same way.

You are right, that you need to implement Generic  model with the same censoring param like in the truncation model.
I think it will be better to review this model in 2-3 weeks (I'll busy during the next week, so i'll continue in one week later).
I'll also want to add tests, because current censored and hurdle models not covered yet.

josef-pkt commented:
I did a brief check on the literature
Stata cpoisson has some references
Cameron Trivedi have a section on hurdle models: They mention that Mullahy in the original paper on modified count models uses the same distribution (family) for the zeros as for the non-zero part, e.g. Poisson. The advantage is that we can test against a single distribution by equality of the two sets of parameters when using the same exog in both parts. (I didn't check yet: CT mention that Mullahy also has a Hausman test as specification test.)
The now more common specification is to use Logit or Probit as a binary model for the zeros.

I haven't checked Winkelman's textbook yet, which is also in the Stata references for cpoisson.

One possibility would be to consider the current Censored models as additional binary models, and allow Hurdle to use either the BinaryCensored or a binary model, Logit, Probit, for the zeros.

(Another obvious extension would be to allow Hurdle models to specify a larger than zero transition point, so we combine a general Censored with a general Truncated count model, i.e. stitch together two count models. I haven't seen this in any reference but it would be just one additional parameter in hurdle models when the general Censored count models are available.)

josef-pkt commented on statsmodels/discrete/count_model.py:
```diff
@@ -0,0 +1,1062 @@
+        llf_main = self.model_main.loglikeobs(params)
+        
+        llf = np.concatenate((llf_main[self.zero_idx],
+            np.log(1 - np.exp(llf_main[self.nonzero_idx]))))
```
as explanation:
`np.exp(llf_main[self.nonzero_idx]))` is the probability for y=0 given x, because the model_main has endog=zeros

josef-pkt commented:
I'm trying to rebase this on current main

merge conflict are very difficult
merge diff wants to convert ZeroInflated model into Truncated Model, so we get many small conflicts
